x-pack/plugins/alerting/server/task_runner/ad_hoc_task_runner.ts

```diff
@@ -0,0 +1,603 @@
+      const {
+        params: { adHocRunParamsId, spaceId },
+        startedAt,
+      } = this.taskInstance;
+
+      const namespace = this.context.spaceIdToNamespace(spaceId);
+
+      this.alertingEventLogger.initialize({
+        context: {
+          savedObjectId: adHocRunParamsId,
+          savedObjectType: AD_HOC_RUN_SAVED_OBJECT_TYPE,
+          spaceId,
+          executionId: this.executionId,
+          taskScheduledAt: this.taskInstance.scheduledAt,
+          ...(namespace ? { namespace } : {}),
+        },
+        runDate: this.runDate,
+        // in the future we might want different types of ad hoc runs (like preview)
+        type: executionType.BACKFILL,
+      });
+
+      let adHocRunData: AdHocRun;
+
+      try {
+        const adHocRunSO: SavedObject<AdHocRunSO> =
+          await this.context.encryptedSavedObjectsClient.getDecryptedAsInternalUser<AdHocRunSO>(
+            AD_HOC_RUN_SAVED_OBJECT_TYPE,
+            adHocRunParamsId,
+            { namespace }
+          );
+
+        adHocRunData = {
+          id: adHocRunSO.id,
+          ...adHocRunSO.attributes,
+          rule: {
+            ...adHocRunSO.attributes.rule,
+            id: adHocRunSO.references[0].id,
+          },
+        };
+      } catch (err) {
+        const errorSource = SavedObjectsErrorHelpers.isNotFoundError(err)
+          ? TaskErrorSource.USER
+          : TaskErrorSource.FRAMEWORK;
+
+        throw createTaskRunError(
+          new ErrorWithReason(RuleExecutionStatusErrorReasons.Decrypt, err),
+          errorSource
+        );
+      }
+
+      const { rule, apiKeyToUse, schedule } = adHocRunData;
+
+      let ruleType: UntypedNormalizedRuleType;
+      try {
+        ruleType = this.ruleTypeRegistry.get(rule.alertTypeId);
+      } catch (err) {
+        throw createTaskRunError(
+          new ErrorWithReason(RuleExecutionStatusErrorReasons.Read, err),
+          TaskErrorSource.FRAMEWORK
+        );
+      }
+
+      this.ruleTypeId = ruleType.id;
+      this.ruleId = rule.id;
+      this.alertingEventLogger.addOrUpdateRuleData({
+        id: rule.id,
+        type: ruleType,
+        name: rule.name,
+        consumer: rule.consumer,
+        revision: rule.revision,
+      });
+
+      try {
+        this.ruleTypeRegistry.ensureRuleTypeEnabled(rule.alertTypeId);
+      } catch (err) {
+        throw createTaskRunError(
+          new ErrorWithReason(RuleExecutionStatusErrorReasons.License, err),
+          TaskErrorSource.USER
+        );
+      }
+
+      let validatedParams: RuleTypeParams;
+      try {
+        validatedParams = validateRuleTypeParams<RuleTypeParams>(
+          rule.params,
+          ruleType.validate.params
+        );
+      } catch (err) {
+        throw createTaskRunError(
+          new ErrorWithReason(RuleExecutionStatusErrorReasons.Validate, err),
+          TaskErrorSource.USER
+        );
+      }
+
+      if (apm.currentTransaction) {
```

Comment:
You're right! We do two calls in the alerting task runner because we don't have all the information in the first call but we do here. Consolidated in https://github.com/elastic/kibana/pull/177640/commits/5cad5288f9ce294714849e3c41648c041e8c7d60